soc/arm/nordic_nrf/nrf53/soc.c

```diff
@@ -137,12 +143,177 @@ bool z_arm_on_enter_cpu_idle(void)
 		suppress_message = true;
 	}
 #endif
+#if defined(CONFIG_SOC_NRF53_RTC_PRETICK) && defined(CONFIG_SOC_NRF5340_CPUNET)
+	if (ok_to_sleep) {
+		extern uint32_t rtc_pretick_rtc1_ticks_to_next_event_get(uint32_t cntr);
+
+		uint32_t rtc1_counter = nrf_rtc_counter_get(NRF_RTC1);
+		uint32_t rtc1_ticks_to_next_event = rtc_pretick_rtc1_ticks_to_next_event_get(rtc1_counter);
+		uint32_t rtc1_pretick_cc_val = (rtc1_counter + rtc1_ticks_to_next_event - 1U) & NRF_RTC_COUNTER_MAX;
+		static bool rtc1_pretick_cc_set_on_time = false;
+
+		if (rtc1_pretick_cc_val != nrf_rtc_cc_get(NRF_RTC1, RTC1_PRETICK_CC_CHAN)) {
+			/* The CC for pretick needs to be updated. */
+			nrf_rtc_cc_set(NRF_RTC1, RTC1_PRETICK_CC_CHAN, rtc1_pretick_cc_val);
+
+			/* Let's check if we updated CC on time so that the CC can take effect. */
+			rtc1_counter = nrf_rtc_counter_get(NRF_RTC1);
+			if (((rtc1_pretick_cc_val - rtc1_counter) & NRF_RTC_COUNTER_MAX) < 2) {
+				/* The COUNTER value is close enough to the expected pretick CC,
+				 * that the pretick event generation is not guaranteed.
+				 */
+				rtc1_pretick_cc_set_on_time = false;
+			} else {
+				/* The written rtc1_pretick_cc is guaranteed to to trigger compare event. */
```

Comment:
There is also a case when next evt was `now+1` and pretick then was set to `now` (the now that was read at the beginning of the function). Second `now` read could progress by 1 tick so `rtc1_pretick_cc_val - rtc1_counter` will give 0xffffff so comment won't be true but it will mean that event just expired so there will be ISR and there is no need for pretick. Am I right or messed it up? If yes then maybe add something like `or user event just expired`.

Reply:
@nordic-krch I believe I'm handling the case you mentioned correctly, please have a look.

Reply:
yes, imo this case is handled correctly. I was only referring to the statement in the comment.

Reply:
comment updated